util/network/src/client_version.rs

```diff
@@ -0,0 +1,473 @@
+
+}
+
+
+fn is_parity(client_id: &str) -> bool {
+	if client_id.starts_with(LEGACY_CLIENT_ID_PREFIX) || client_id.starts_with(PARITY_CLIENT_ID_PREFIX) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+// Parse known parity formats.
+//
+// This is really not robust: parse four arguments and
+// allow for an extra argument between identifier and
+// version
+// TODO implement a better logic
+fn parse_parity_format(client_version: &str) -> Result<ParityClientData, ()> {
+	let tokens: Vec<&str> = client_version.split("/").collect();
+
+
+	// Basically strip leading 'v'
+	if let Some(version_number) = &get_number_from_version(tokens[1]) {
+		if tokens.len() < 4 {
+			return Err(());
+		}
+
+		return Ok(
+			ParityClientData {
+				name: tokens[0].to_string(),
+				variant: None,
+				semver: Version::parse(version_number).unwrap(),
+				os: tokens[2].to_string(),
+				compiler: tokens[3].to_string(),
+			}
+		);
+	} else if let Some(version_number) = &get_number_from_version(tokens[2]) {
+		if tokens.len() < 5 {
+			return Err(());
+		}
+
+		return Ok(
+			ParityClientData {
+				name: tokens[0].to_string(),
+				variant: Some(tokens[1].to_string()),
+				semver: Version::parse(version_number).unwrap(),
+				os: tokens[3].to_string(),
+				compiler: tokens[4].to_string(),
+			}
+		);
+	} else {
+		return Err(());
+	}
+}
+
+// Parses a version string and returns the corresponding
+// ClientVersion. Only Parity clients are destructured right now.
+// The parsing for parity may still fail, in which case return an Other with
+// the original version string. TryFrom would be a better trait to implement.
+
+impl From<&str> for ClientVersion {
+	fn from(client_version: &str) -> Self {
+		if !is_parity(client_version) {
+			return ClientVersion::Other(client_version.to_string());
+		}
+
+		if let Ok(data) = parse_parity_format(client_version) {
+			ClientVersion::ParityClient(data)
+		} else {
+			ClientVersion::ParityUnknownFormat(client_version.to_string())
+		}
+	}
+}
+
+impl From<String> for ClientVersion {
+	fn from(client_version: String) -> Self{
+		ClientVersion::from(client_version.as_ref())
+	}
+}
+
+fn format_parity_version_string(client_version: &ParityClientData, f: &mut fmt::Formatter) -> std::fmt::Result {
+	let name = client_version.name();
+	let semver = client_version.semver();
+	let os = client_version.os();
+	let compiler = client_version.compiler();
+
+	match client_version.variant() {
+		None => write!(f, "{}/v{}/{}/{}", name, semver, os, compiler),
+		Some(variant) => write!(f, "{}/{}/v{}/{}/{}", name, variant, semver, os, compiler),
+	}
+}
+
+impl fmt::Display for ClientVersion {
+	fn fmt(&self, f: &mut fmt::Formatter) -> std::fmt::Result {
+		match self {
+			ClientVersion::ParityClient(data) => format_parity_version_string(data, f),
+			ClientVersion::ParityUnknownFormat(id) => write!(f, "{}", id),
+			ClientVersion::Other(id) => write!(f, "{}", id)
+		}
+	}
+}
+
+fn get_number_from_version(version: &str) -> Option<String> {
```

Comment:
Do we really need to allocate new `String` here? Maybe `&str` would do?

Reply:
Sure, good point.